src/thor/chinese_postman_action.cc

```diff
@@ -0,0 +1,469 @@
+#include <algorithm>
+
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/properties.hpp>
+#include <boost/multi_array.hpp>
+#include <boost/multi_array/subarray.hpp>
+
+#include "Hungarian.h"
+
+#include "loki/search.h"
+#include "midgard/util.h"
+#include "sif/costconstants.h"
+#include "sif/recost.h"
+#include "thor/chinese_postman_graph.h"
+#include "thor/costmatrix.h"
+#include "thor/worker.h"
+#include "tyr/serializers.h"
+
+using namespace valhalla::baldr;
+using namespace valhalla::midgard;
+using namespace valhalla::sif;
+
+namespace valhalla {
+namespace thor {
+
+midgard::PointLL to_ll(const valhalla::Location& l) {
+  return midgard::PointLL{l.ll().lng(), l.ll().lat()};
+}
+
+// Return the index of an edge compared to the path_edge from a location. Assuming the path_edge is
+// ordered by the best.
+int get_node_candidate_index(const valhalla::Location& location,
+                             const GraphId& edge_id,
+                             double* percent_along) {
+  // see if this edge is an edge candidate for this location
+  for (int i = 0; i < location.correlation().edges_size(); ++i) {
+    const auto& e = location.correlation().edges(i);
+    if (e.graph_id() == edge_id) {
+      // TODO: this will always be 0, this is definitely a bug
+      *percent_along = int(e.percent_along());
```

Comment:
@ismailsunni i started cleaning up this source file a bit, simplifying things, refactoring a bit, adding some protections, and updating some comments but one thing i foudn that looks like a bug to me was this. previously you had passed `percent_along` as `int*` which forced the floating point number (always between 0-1) to be truncated, most likely to 0. if i let this thing stay as a floating point number lots of tests fail. is this really the behavior we wanted here? i saw down below where you check it to see whether its closer to 0 or closer to 1, whcih made me think this is not the behavior we really wanted. let me know what you think!

Reply:
Hi @kevinkreiser you are correct, it's not supposed to be an `int*`. I have tried your latest commit and removed the int conversion. It indeed failed a lot of tests, but actually only 2 (`TestChinesePostmanCouldNotFindingMatchingEdge` and `TestChinesePostmanMiddleEdge`). Both are throwing the same error: `Could not find matching edge candidate`. I have time today to debug it, hopefully, I got something.

Reply:
awesome! let me know what you find. i'll keep working on cleaning up the code on my end too! there are a lot of ID to string conversions and ordered maps that I want to switch to ID hash maps. i'll get to that probably this weekend. sorry again this is taking so long but i finally feel like i have just enough time to make progress on it!

Reply:
hi @kevinkreiser 
It seems the issue is related to how the starting and end points handle the percent_along. I have checked that if I use double for the percent along, all tests are ok except when the one that the starting or endpoints are not in the node.
In this case, the algorithm seems ok, but how I handle the percent along is not ok if the percent_along is not 0 or 1.

The previous success test was a lucky case.

I have disabled and commented out on the failed test. I hope I can do more debugging next week.

Reply:
sounds great! i might be able to take a look as well, ill let you know. i have some more changes coming too to speed up the cp graph a bit. its great to make progress on getting this ready for merge. thanks again for all the work!